libdrgn/iterators.h

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#ifndef DRGN_ITERATORS_H
+#define DRGN_ITERATORS_H
+
+#include <stddef.h>
+#include <stdint.h>
+#include "drgn.h"
+#include "vector.h"
+
+struct radix_tree_iter_frame {
+	struct drgn_object slots;
+	uint64_t index;
+	uint64_t shift;
+	uint64_t loop_index;
+};
+
+DEFINE_VECTOR_TYPE(frames, struct radix_tree_iter_frame)
+
+typedef struct radix_tree_iter_entry {
+	uint64_t index;
+	struct drgn_object node;
+} radix_tree_iter_entry_t;
+
+struct radix_tree_iter {
+	bool started;
+	struct drgn_object root;
+	// Current value to be yielded
+	struct radix_tree_iter_entry entry;
+	// We need this for later initialization of `drgn_object`s
+	struct drgn_program *prog;
+	// Frames to keep track of generator state
+	struct frames frames;
+	// One-time setup values that are persistent
+	uint64_t RADIX_TREE_INTERNAL_NODE;
+	uint64_t RADIX_TREE_MAP_MASK;
+	struct drgn_qualified_type node_type;
+};
+
+struct drgn_error *radix_tree_iter_init(struct drgn_program *prog,
+					const struct drgn_object *root,
+					struct radix_tree_iter *ret);
+
+struct drgn_error *radix_tree_iter_next(struct radix_tree_iter *iter,
+					struct radix_tree_iter_entry **ret);
```

Comment:
Just to make sure I understand, the semantics of `ret` are:

* It's a pointer to a structure embedded in `iter`.
* The caller is free to do what it wishes with it as long as the `drgn_object` remains initialized, but
* It will be overwritten the next time that this is called

At first, I was going to suggest having the caller instead pass in a `struct drgn_object *` that would be set to the result, but then it's somewhat difficult to follow the general rule that a result object is not modified if the function returns an error. So, I think this is fine, but please document it. (It might be better to return a `struct drgn_object *` and the `uint64_t` index as separate out parameters, though).

Reply:
Yes, precisely.